callout/src/main/java/com/google/apigee/JavaCalloutRemotePolicyExecution.java

```diff
@@ -0,0 +1,123 @@
+import com.apigee.flow.message.MessageContext;
+import com.google.apigee.Execute.ExecutionResult.Action;
+import com.google.apigee.ProtoMessageBuilders.ExecutionContextProtoMessageBuilder;
+import com.google.apigee.ProtoMessageBuilders.MessageContextProtoMessageBuilder;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * Java Callout demonstrating execution of a Java Callout or Apigee Policy on remote HTTP Server.
+ */
+public class JavaCalloutRemotePolicyExecution implements Execution {
+
+  private final String FLOW_VARIABLE_KEY = "Example";
+  private final String REMOTE_SERVER_URL_PROPERTY_NAME = "remote_execution_url";
+  private final RemotePolicyExecutionHandler remotePolicyExecutionHandler;
+  private Map<String, String> properties;
+
+  public JavaCalloutRemotePolicyExecution(Map<String, String> properties) {
+    this(properties, new RemotePolicyExecutionHandler());
+  }
+
+  public JavaCalloutRemotePolicyExecution(
+      Map<String, String> properties, RemotePolicyExecutionHandler remotePolicyExecutionHandler) {
+    this.properties = properties;
+    this.remotePolicyExecutionHandler = remotePolicyExecutionHandler;
+  }
+
+  /**
+   * Constructs a Protocol Buffer Message using the {@link MessageContext} and {@link
+   * ExecutionContext} objects and sends them over to the remote HTTP server for execution. HTTP
+   * Server URL is retrieved from a flow variable.
+   *
+   * @param messageContext Object allowing access to entities inside the flow
+   * @param executionContext Object allowing access to proxy execution context
+   * @return A successful execution after response is received from HTTP Server
+   */
+  public ExecutionResult execute(MessageContext messageContext, ExecutionContext executionContext) {
+    try {
+      // URL of the HTTP Server endpoint set in a flow variable.
+      String serverUrl = this.properties.get(REMOTE_SERVER_URL_PROPERTY_NAME);
+      Execute.Execution execution =
+          Execute.Execution.newBuilder()
+              .setExecutionContext(
+                  ExecutionContextProtoMessageBuilder.buildExecutionContextProto(executionContext))
+              .setMessageContext(
+                  MessageContextProtoMessageBuilder.buildMessageContextProto(messageContext))
+              .build();
+      Execute.Execution remoteExecutionResult =
+          remotePolicyExecutionHandler.sendRemoteHttpServerRequest(execution, serverUrl);
+      ExecutionResult executionResult =
+          remoteExecutionResult.hasExecutionResult()
+              ? extractExecutionResult(remoteExecutionResult.getExecutionResult())
+              : ExecutionResult.SUCCESS;
+      if (executionResult.getAction() == com.apigee.flow.execution.Action.CONTINUE) {
+        messageContext
+            .getMessage()
+            .setContent(extractFlowVariable(remoteExecutionResult, FLOW_VARIABLE_KEY));
+      }
+      return executionResult;
+    } catch (Throwable throwable) {
```

Comment:
is there a way to log this?

Reply:
A nice way would be to import the BaseClass I made and call the `setExceptionVariable` method which stores the exception, message and stacktrace in flow variables. Unfortunately, since the BaseClass doesn't exist in a maven repository yet, we'd have to add a local system dependency to import it and bundle the jar to be uploaded when deploying the api proxy.

An easier way (for now, until the BaseClass is uploaded to a maven repo) would be to store the exception message in a flow variable but it wouldn't include the stacktrace and other neat info unless we were to copy & paste that code in.

Reply:
The base class way sounds good. Could you leave a comment in the code as a TODO documenting this?

Maybe we can also use a logger so the same messages are printed to the console or whatever log files that get created.

Reply:
how about adding a TODO?

Reply:
Added TODO. I don't think customers can actually see anything printed to console or access any log files for callouts executed on Apigee MPs. The baseclass also has a `log` method that stores log messages in flow variables which customers can see.